immuneML/workflows/instructions/clustering/ClusteringInstruction.py

```diff
@@ -0,0 +1,204 @@
+        encoded_dataset = self.encode(unit, result_path / "encoded_dataset")
+        encoded_dataset.name = unit.dataset.name
+        if unit.dim_red_before_clustering:
+            self._dim_reduce(unit, encoded_dataset)
+
+        unit.clustering_method.fit(encoded_dataset.encoded_data)
+
+        labels_true = None
+        if unit.true_labels_path is not None and unit.true_labels_path.is_file():
+            try:
+                labels_true = genfromtxt(unit.true_labels_path, dtype=int, delimiter=',')
+            except:
+                print_log("Problem getting true_labels_path file\nCheck the file is in the right format(CSV, 1 line)")
+        self.calculate_scores(key, encoded_dataset.encoded_data.examples, unit.clustering_method.model.labels_, labels_true)
+
+        if not unit.dim_red_before_clustering:
+            self._dim_reduce(unit, encoded_dataset)
+
+        processed_dataset = self.add_label(encoded_dataset, unit.clustering_method.model.labels_, result_path / "dataset_clustered")
+
+        unit.report.dataset = processed_dataset
+        unit.report.method = unit.clustering_method
+        unit.report.result_path = result_path / "report"
+        unit.report.number_of_processes = unit.number_of_processes
+        print_log("Report generation started...", include_datetime=True)
+        report_result = unit.report.generate_report()
+        print_log("Report generating finished.", include_datetime=True)
+        return report_result
+
+    def encode(self, unit: ClusteringUnit, result_path: Path) -> Dataset:
+        if unit.encoder is not None:
+            encoded_dataset = DataEncoder.run(DataEncoderParams(dataset=unit.dataset, encoder=unit.encoder,
+                                                                encoder_params=EncoderParams(result_path=result_path,
+                                                                                             label_config=unit.label_config,
+                                                                                             pool_size=unit.number_of_processes,
+                                                                                             filename="encoded_dataset.pkl",
+                                                                                             learn_model=True,
+                                                                                             encode_labels=unit.label_config is not None),
+                                                                ))
+        else:
+            encoded_dataset = unit.dataset
+        return encoded_dataset
+
+    def _dim_reduce(self, unit: ClusteringUnit, dataset):
+        if unit.dimensionality_reduction is not None:
+            unit.dimensionality_reduction.fit(dataset.encoded_data)
+            unit.dimensionality_reduction.transform(dataset.encoded_data)
+
+    def calculate_scores(self, key, data, labels_pred, labels_true):
+        if self.state.clustering_scores is None:
```

Comment:
make sure that this is configurable, split maybe into functions or compute automatically, it is not necessary to always compute all scores?